### PR TITLE
Merge dev-2018-cssfix to dev-2018

### LINKED ISSUE #29: GCS 2018-2019 Remodel
Create a more flexible GCS UI which includes the current GCS with additional features.
- improved Mission Status Container
- improved Vehicle Status Container
- more interaction with Console Container (and less spammy than SLO GCS)
- consistent code style between developers using linter
   - use consistent text editor ([Atom](https://atom.io) and [Sublime 3](https://www.sublimetext.com) work well)

Node modules to consider: [React](https://reactjs.org), [Material UI](https://material-ui.com)
Remodeled GCS in development: see [Luis729/gcs-2018](https://github.com/luis729/gcs-2018)

**Note**: Not considering communication backend for this yet. That should be easy to implement with renderer notifications.

### PR DESCRIPTION
**Pull request by Luis729**
List of changes:

1. Map caching works. It currently automatically caches data as long as the `cache` configuration in `resources/config.json` is enabled.

2. Containers are fixed and do not overlap over one another. Design of UI is slicker now, with its dark theme design.

3. Map supports Xbee library through the `serialport` module. We keep the module at v6 to keep it compatible with Electron 3.

4. Removed support for Protobuf to move GCS to JSON protocol communication

Things to work on:

1. There are a few issues still going with the UI in various platforms. An example is that it does not close when closing the window in Windows or Linux.

2. Make caching manual, so that the user can select the area to cache to memory.

Fixes #29, #31, #32, #33

### CODE REVIEW

**Comment by Luis729** [comment]
While I resolve conflicts, I would like to verify that `dev-2018-cssfix` works on all platforms before merging.

The main concern is that `serialport` module does not crash the entire program when imported.

To test this, simply insert the following line into `src/renderer/index.js`

```js
import SerialPort from 'serialport';
```

If the program still runs, then `serialport` is successful in that certain OS.

- [x] macOS
- [x] Windows
- [x] Linux

**Review by maxwey — COMMENTED** [review]
Runs on Mac. 

What do I need to do to ensure that it runs on linux (how do I check that it opens the correct port; or @Luis729 are you going to check that?)

File: resources/mac-address.json
Comment (maxwey) [inline resources/mac-address.json]: please make this a list of possible Mac addresses. 

The code should be able to go through all the possible Mac addresses that the GCS can be and pick the first that works.

Since vehicles will likely need this information too, we should decide what the radios we will use are and make it run with any of the possible radios.

File: src/renderer/log/log.css
Comment (maxwey) [inline src/renderer/log/log.css]: honestly, not a fan of the black log background... but I guess if we're going with a dark scheme...

Maybe make it a little off-black? like #141414? or something?

File: src/renderer/map/Map.js
Comment (maxwey) [inline src/renderer/map/Map.js:42]: please move field variables to the top of the class definition.

File: src/renderer/map/Map.js
Comment (maxwey) [inline src/renderer/map/Map.js:24]: what is the purpose of this logic? and why can we not move this inside the constructor?

File: src/renderer/map/Map.js
Comment (maxwey) [inline src/renderer/map/Map.js:14]: why are these not field variables?

File: src/renderer/mission/Mission.js
Comment (maxwey) [inline src/renderer/mission/Mission.js]: what is the convention for using _ in method names?

File: src/renderer/vehicle/Vehicle.js
Comment (maxwey) [inline src/renderer/vehicle/Vehicle.js]: move field value  to top of class

File: test/updateMessages.test.js
Comment (maxwey) [inline test/updateMessages.test.js]: how are these tests run? Does npm run them with `npm test`? or some other npm command?

File: src/renderer/log/Log.js
Comment (maxwey) [inline src/renderer/log/Log.js]: field variable location

**Review by maxwey — CHANGES_REQUESTED** [review]
Sorry I found a couple more things.

Specifically for the class organization, they should be specified in the following order:
1. all static field variables
2. remaining (non static) variables
3. all the field variable lambdas (`myFunc = () => { ... }`)
4. the contractor or similar to constructor functions
5. all the normal methods

File: src/main/index.js
Comment (Luis729) [reply src/main/index.js:94]: I would like to remove this and have a better option (like a popup window) for user to add new locations instead in the future.

Besides editing a config file sounds tacky. If the user edits the config file, assuming the user knows what to do, the menu will not update with the new location (or deleted locations) anyway.

Also I think I had issues maintaining location file when I turned the application to a Desktop App (gotta check if it can compile to an app all the time, right?)

Comment (Luis729) [reply src/main/index.js:94]: Also can you check if it runs on Linux? Just go to `src/renderer/index.js` and try to import `serialport` into the code. If serialport doesnt crash the whole thing it should work. No need to test with specific Xbee ports since I have done that. As long as you know the specific port name for Linux you should be fine. Like Windows has the `COM5` and macOS has the `/dev/tty.portETC`, not sure for Linux

File: resources/mac-address.json
Comment (Luis729) [reply resources/mac-address.json]: I do not think we need a list of possible MAC addresses. Each platform will have a specific Xbee that will be mounted on the vehicle, as well as extra Xbees for devleopment (in the case of GCS, we will have one of the 6 new Xbees be the field test Xbee).

Now to have everyone know what platforms have what Xbees, here's what I propose:
1. Put a table of everyone's MAC addresses on the Wiki
2. Have a file called `mac-addresses.txt` in the root of [NGCP/Communication](https://github.com/NGCP/Communication). I remember other languages struggle to process JSON files; but I mean if they have a JSON to class converter it shouldn't be too bad right?

Comment (Luis729) [reply resources/mac-address.json]: Remind me to add your rule on class organization to the wiki somewhere.

> Sorry I found a couple more things.
> 
> Specifically for the class organization, they should be specified in the following order:
> 
> 1. all static field variables
> 2. remaining (non static) variables
> 3. all the field variable lambdas (`myFunc = () => { ... }`)
> 4. the contractor or similar to constructor functions
> 5. all the normal methods

Comment (Luis729) [reply resources/mac-address.json]: Also I remember why I put some variables after certain functions. It's cause of the linter... You think I should just disable the rule? See image below:

screen shot 2019-02-08 at 4 44 12 pm

File: src/renderer/mission/Mission.js
Comment (Luis729) [reply src/renderer/mission/Mission.js]: I see people doing _name when putting the function into a parameter.

```js
<Table
  width={width}
  height={height}
  headerHeight={40}
  rowHeight={this.heightCache.rowHeight}
  rowCount={this.state.missions.length}
  rowGetter={this._rowGetter}
  onRowClick={this._onRowClick}
>
```

Seems to be a standard of some sort when it comes to naming functions. Consider them to be like private functions (people call them callbacks)

File: test/updateMessages.test.js
Comment (Luis729) [reply test/updateMessages.test.js]: I plan to implement tests like how you guys do (mocha).

These "tests" are more of fixtures I throw into the GUI to see how it visually looks. Like how you did it last year

Comment (Luis729) [reply test/updateMessages.test.js]: I think I will be adding mocha testing to the branch before fully merging to `dev-2018`, as well as creating a separate folder called `fixtures` and adding that in.

We can change the name `devMode` in the `config.json` to `fixtures` to be more clear about that.

File: src/renderer/log/log.css
Comment (Luis729) [reply src/renderer/log/log.css]: Resolved in https://github.com/NGCP/GCS/commit/7281b09d05d28cd92286e87185713d2b03d56da1

File: src/renderer/mission/Mission.js
Comment (Luis729) [reply src/renderer/mission/Mission.js]: Resolved in https://github.com/NGCP/GCS/commit/7281b09d05d28cd92286e87185713d2b03d56da1

File: test/updateMessages.test.js
Comment (Luis729) [reply test/updateMessages.test.js]: Resolved in https://github.com/NGCP/GCS/commit/7281b09d05d28cd92286e87185713d2b03d56da1

File: src/main/index.js
Comment (maxwey) [reply src/main/index.js:94]: I don't have an xbee on me or at home. Does that matter?

File: resources/mac-address.json
Comment (maxwey) [reply resources/mac-address.json]: I don't like having method and variables mixed. I'm all for disabling that rule, unless there's like a solid reason why it needs to be in that order.

So why I want a list of  possible MAC addresses: the xbee im gonna use for GCS is not the XBee your gonna use. If we don't want to manipulate config files (which I agree with you there, we don't want to) then we need to either add ui element to allow the user to input the Mac address, or we detect the xbee radio's Mac address, or we establish a list of the possible xbee gcs radios (the ones you will use for gcs and the ones I will use for gcs) and then add them all to the config file and have it figure out the one it needs.

File: src/main/index.js
Comment (Luis729) [reply src/main/index.js:94]: No. Just see if the program runs with `import 'serialport';` in it

File: resources/mac-address.json
Comment (Luis729) [reply resources/mac-address.json]: Sure. We can do it like that for development purpose. I assume for all other platforms we will stick with one value

Comment (Luis729) [reply resources/mac-address.json]: For the rule, since it enacts itself on React components only, I will modify the rule to make it adapt to your rules. I wonder if there's a rule that will enforce structuring a class the way you do it. Otherwise we will have developers that will be creating them wrong.

Comment (Luis729) [reply resources/mac-address.json]: Updated rules and mac-addresses

File: src/renderer/map/Map.js
Comment (Luis729) [reply src/renderer/map/Map.js:14]: Each vehicle will have a certain icon. For sake of example we will use UAV.

UAV that is connected will have a black border around it.
UAV that is disconnected with have a red border around it.

We can add others as we go but that's kinda where we are right now.

Comment (Luis729) [reply src/renderer/map/Map.js:14]: We will also have the image to have the format of the vehicle platform.

We have uav.png for all UAVs. We plan to create a multirotor.png for all multirotor vehicles. These will have black borders.

All vehicles with red borders will have _red appended to them. For example: uav_red.png, multirotor_red.png

Software will automatically assign images to the vehicles hence.

File: src/renderer/map/Map.js
Comment (Luis729) [reply src/renderer/map/Map.js:24]: Just made another commit that readded locations.json.

Just in that crazy case that locations.json is not there, the program will not crash. Also I cannot put this inside the class since the state variable depends on this variable, and is declared in the constructor (in a weird way through ES7)

I would leave that code there though

Comment (Luis729) [reply src/renderer/map/Map.js:24]: If you remember from last year, we needed a default start location for the map. From the code, we just set it to 0,0. If there's a set startLocation and that startLocation is valid, then the map will start at that location instead.

Usually geolocation will come and it will pan at your location a few seconds later.

File: resources/mac-address.json
Comment (Luis729) [reply resources/mac-address.json]: I realized that we don't actually need a list of MAC addresses for GCS, since we aren't sending messages to ourselves. It is other platforms that will be sending messages to us, so that will be them that needs those MAC addresses.

When that case happens, either they can implement having a list of GCS MAC addresses, or just have them put the Xbee MAC address we are using for field test (I prefer the latter).

I'll be removing the gcs field in the json file in a future commit, as it is not needed.

**Review by maxwey — APPROVED** [review]

**Comment by maxwey** [comment]
Runs on Linux.

**Pull request merged**
